File: mongodb_schema/game_odds.js
db = db.getSiblingDB('fox_cub');

db.createCollection("game_odds", {
    validator: {
        $jsonSchema: {
            bsonType: "object",
            required: [ "home_team", "away_team", "odds", "tournament", "event_date" ],
            properties: {
                home_team: {
                    bsonType: "string",
                    description: "must be a string and is required"
                },
                away_team: {
                    bsonType: "string",
                    description: "must be a string and is required"
                },
                odds: {
                    bsonType: "array",
                    description: "must be a array and is required"
                },
                event_date: {
                    bsonType: "date",
                    description: "must be a date and is required"
                },
                tournament: {
                    bsonType: "string",
                    description: "must be a string and is required"
                },
            }
        }
    }
})
 

File: src/main/scala/entities/GameOdds.scala
package fox.cub.model

import io.vertx.lang.scala.json.Json
import io.vertx.core.json.JsonObject

import fox.cub.internals.QueryEvent
import fox.cub.utils.Utils.getUTCdate

object GameOdds {
    private val Collection = "game_odds"

    /**
     * Get list of not expired game odds
    */
    def getRecent(tournamentId: String): QueryEvent = {
        val dateFilter = Json.obj(("$gt", Json.obj(("$date", getUTCdate()))))

        val aggMatch = Json.obj(
            ("$match", Json.obj(
                ("tournament", tournamentId),
                ("event_date", dateFilter)
            )))

        val teamsToObjects =  Json.obj(("$addFields", Json.obj(
                ("home_team", Json.obj(("$toObjectId", "$home_team"))),
                ("away_team", Json.obj(("$toObjectId", "$away_team")))
            )))

        val joinHomeTeam = Json.obj(("$lookup", Json.obj(
            ("from", "team"),
            ("localField", "home_team"),
            ("foreignField", "_id"),
            ("as", "home_team")
            )))

        val joinAwayTeam = Json.obj(("$lookup", Json.obj(
            ("from", "team"),
            ("localField", "away_team"),
            ("foreignField", "_id"),
            ("as", "away_team")
            )))

        var pipeline = Json.arr(aggMatch, teamsToObjects, joinHomeTeam, joinAwayTeam)
        // cursor with the default batch size
        var cursor = Json.obj()
        val query = new JsonObject().put("aggregate", Collection).
            put("pipeline", pipeline).put("cursor", cursor)

        QueryEvent("aggregate", query)
    }

}
